src/preprocessing/passes/non_clausal_simp_sat.cpp

```diff
@@ -0,0 +1,483 @@
+ **
+ ** \brief Non-clausal simplification preprocessing pass.
+ **
+ ** Run the nonclausal solver and try to solve all assigned theory literals.
+ **/
+
+ #include "preprocessing/passes/non_clausal_simp_sat.h"
+ #include <cryptominisat5/cryptominisat.h>
+
+ #include "prop/registrar.h"
+ #include "prop/cnf_stream.h"
+ #include "prop/sat_solver_factory.h"
+
+ #include "proof/proof_manager.h"
+
+ #include "context/cdo.h"
+ #include "options/proof_options.h"
+ #include "smt/smt_statistics_registry.h"
+ #include "theory/theory_model.h"
+#include <ctime>
+
+#include <vector>
+
+using namespace CVC4;
+using namespace CVC4::theory;
+
+namespace CVC4 {
+namespace preprocessing {
+namespace passes {
+
+
+/* -------------------------------------------------------------------------- */
+
+NonClausalSimpSAT::NonClausalSimpSAT(PreprocessingPassContext* preprocContext)
+    : PreprocessingPass(preprocContext, "non-clausal-simp-sat")
+{
+}
+
+PreprocessingPassResult NonClausalSimpSAT::applyInternal(
+    AssertionPipeline* assertionsToPreprocess)
+{
+
+  Assert(!options::unsatCores() && !options::fewerPreprocessingHoles());
+
+  d_preprocContext->spendResource(options::preprocessStep());
+
+
+  // Convert the original formula into a boolean formula and solve it
+  // with cryptominisat
+  SatSolver* d_satSolver = SatSolverFactory::createCryptoMinisat(
+                smtStatisticsRegistry(), "non_clausal_simp_solver");
+  CVC4::prop::NullRegistrar d_registrar;
+  context::Context d_context;
+  CVC4::prop::TseitinCnfStream d_cnfStream (d_satSolver, &d_registrar,
+                                            &d_context, true,
+                                            "toCNF-non-clausal-simp");
+
+  Trace("non-clausal-simplify-sat") << "asserting to sat solver" << std::endl;
+
+  unsigned substs_index = d_preprocContext->getSubstitutionsIndex();
+  for (size_t i = 0, size = assertionsToPreprocess->size(); i < size; ++i)
+  {
+    Assert(Rewriter::rewrite((*assertionsToPreprocess)[i])
+              == (*assertionsToPreprocess)[i]);
+       // Don't reprocess substitutions
+       if (substs_index > 0 && i == substs_index)
+       {
+         continue;
+       }
+       Trace("non-clausal-simplify-sat")
+           << "asserting " << (*assertionsToPreprocess)[i] << std::endl;
+       d_cnfStream.convertAndAssert((*assertionsToPreprocess)[i], false, false,
+                                    RULE_GIVEN);
+  }
+
+
+
+  SatValue result = d_satSolver->solve();
+
+  if (result==SAT_VALUE_FALSE){
+    // If in conflict, just return false
+    Trace("non-clausal-simplify-sat")
+        << "conflict in non-clausal propagation" << std::endl;
+    Assert(!options::unsatCores() && !options::fewerPreprocessingHoles());
+    assertionsToPreprocess->clear();
+    Node n = NodeManager::currentNM()->mkConst<bool>(false);
+    assertionsToPreprocess->push_back(n);
+    PROOF(ProofManager::currentPM()->addDependence(n, Node::null()));
+    delete d_satSolver;
```

Comment:
Take a look at [`std::unique_ptr`](https://en.cppreference.com/w/cpp/memory/unique_ptr) to for a better way of managing memory without explicit `delete`s.